core/src/main/java/io/grpc/ProxyParameters.java

```diff
@@ -24,13 +24,19 @@
 /**
  * Used to express the result of a proxy lookup.
  */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/5113")
 public final class ProxyParameters {
-  public final InetSocketAddress proxyAddress;
-  @Nullable public final String username;
-  @Nullable public final String password;
 
-  /** Creates an instance. */
-  public ProxyParameters(
+  private final InetSocketAddress proxyAddress;
```

Comment:
This seems like it needs to be a SocketAddress, rather than an InetSocketAddress.  If there was a UDS http proxy, it wouldn't be able to use this.

Reply:
Ok, after reading this https://github.com/grpc/grpc-java/issues/1539 I am starting to understand...

Reply:
Addressed too

Reply:
:-/ It should have stayed a InetSocketAddress, since this is for CONNECT and CONNECT requires IP addresses.

Reply:
@ejona86 when it was an internal detail, yeah, but now that it is becomign public, it should be based on the transport.  A different transport may not use InetSocketAddress.

Reply:
> ... CONNECT requires IP addresses

But, technically the connection from client to the proxy could still be over UDS and the contents sent use the correct HTTP CONNECT semantics (with a destination IP address), so the connection from the proxy to server being over TPC right?

> ... it should be based on the transport. A different transport may not use InetSocketAddress.

Not only in the transport, given that the proxy is an external entity. 

I think the key question is: is there any HTTP proxy implementation that supports UDS incoming connections? So far, I couldn't find any

Reply:
Oh, right, this is the proxy's address. Okay, fair, it can be a `SocketAddress` in some ways. But this will cause havoc with #3085 since a NameResolver would no longer know which address types it might return, as ProxyDetector chooses this proxy address. This wasn't a problem before because proxy selection would only happen for addresses that could have been InetSocketAddress. (That is to say, a NameResolver that only cares about UnixDomainSocketAddress wouldn't even need to consult the ProxyDetector because a proxy will always use inet addresses to connect to the end host.) So even we want to support HTTP proxies for UDS incoming connections, I don't know how we could support it because there's no real way to choose a transport that supports UDS at that point. Even telling the ProxyDetector the permitted address types doesn't really solve it because the dns resolver wouldn't support UDS.

But I guess we can just have a "don't do that" policy. `java.net.Proxy` does have a `SocketAddress`, although it also is "extensible" as there could be new proxy types added in the future.

Maybe we can just say that `ProxyDetector.proxyFor()` should return the same `SocketAddress` type as provided. We'd add a new method in the future if we needed something more advanced.

I really didn't want to get into all this... The API still needs work and we agree it needs work, but there will be very few users of this part of the API (they will receive/pass ProxyParameters, but they won't really create or inspect them).

Reply:
I think we can punt on this if ProxyDetector becomes an abstract class.  I'm not certain that we can enforce the socket types being the same, since InetSocketAddress is nonfinal.